Add single-file app's dir to NATIVE_DLL_SEARCH_DIRECTORIES

mateoatr opened the pull request:
Fixes #42772. I took the easy path here - whenever the app is single-file, add the exe's dir to `NATIVE_DLL_SEARCH_DIRECTORIES`. Additionally, if the user sets `-p:IncludeNativeLibrariesForSelfExtract=true`, add the extraction path to the search directories.

Commit:
Add single-file app's dir to native search directories

vitek-karas commented:
I think the logic should be:
* If this is a single-file - add the bundle's directory to the native search paths
* If we extracted anything - add the extraction directory to the native search paths

This should specifically cover the 3.1 backward compat mode (I know it works today on Windows, but it won't work on Linux ).

We obviously need tests for this - we can start with validating the content of the `NATIVE_DLL_SEARCH_DIRECTORIES` as that is easy enough to do in the tests.

elinor-fung commented on src/installer/corehost/cli/bundle/header.h:
```diff
@@ -41,7 +41,8 @@ namespace bundle
         none = 0,
-        netcoreapp3_compat_mode = 1
+        netcoreapp3_compat_mode = 1,
+        needs_native_libraries_extraction = 2,
```
Do we really need a separate flag in the manifest for this? It seems like we'd be stuck with this oddly tailored flag (unless we rev the bundle version).

Could we just add the extraction path as long as it exists (we extract something)? I know we'd end up adding it unnecessarily in some cases (extracting symbols but not native libraries for example), but I don't think that is particularly risky.

elinor-fung commented on src/installer/corehost/cli/hostpolicy/deps_resolver.cpp:
```diff
@@ -876,6 +876,19 @@ bool deps_resolver_t::resolve_probe_dirs(
+        // If native components are self-extracted, add the extraction path too.
+        if (bundle->needs_native_libraries_extraction())
+        {
+            add_unique_path(asset_type, bundle->extraction_path(), &items, output, &non_serviced, core_servicing);
```
Should this check that the extraction path is non-empty?

mateoatr replied:
I just changed the if condition to check that the directory exists, I don't see any reason why we should also check if it's empty.

elinor-fung reviewed (commented):
Tests?

elinor-fung commented:
> This should specifically cover the 3.1 backward compat mode (I know it works today on Windows, but it won't work on Linux ).

What is broken with 3.1 compat mode on Linux for this? It should (with the fix for actually extracting) be able to search for native libraries based on the managed assembly's path?

vitek-karas commented:
You're right that it will basically work - but as you pointed out - the design called for setting the extraction path as one of the native search directories - which this should fix. There are still potential scenarios (plugins) where not setting it will break the app (mostly on Linux), but I agree that they're not that common.

Commit:
PR feedback

mateoatr commented:
I removed `needs_native_libraries_extraction` from the manifest flags and changed the logic so that whenever the extraction dir exists, we add it to the native search paths. 
For testing I'm just validating that we put the bundle's dir and the extraction dir in `NATIVE_DLL_SEARCH_DIRECTORIES`.

vitek-karas commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/SingleFileApiTests.cs:
```diff
@@ -67,6 +68,23 @@ public void AppContext_Deps_Files_Bundled_Self_Contained()
+        public void AppContext_Native_Search_Dirs_Contains_Bundle_And_Extraction_Dirs()
+        {
+            var fixture = sharedTestState.TestFixture.Copy();
+            Bundler bundler = BundleHelper.BundleApp(fixture, out string singleFile, BundleOptions.BundleNativeBinaries);
```
Please add another test without any extraction - to validate that the logic works correctly

vitek-karas replied:
Also once we merge this with the other PR the tests will need to change

Commit:
Add test case for non-extracting single-file

elinor-fung commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/SingleFileApiTests.cs:
```diff
@@ -67,6 +68,42 @@ public void AppContext_Deps_Files_Bundled_Self_Contained()
+        public void AppContext_Native_Search_Dirs_Contains_Bundle_And_Extraction_Dirs()
+        {
+            var fixture = sharedTestState.TestFixture.Copy();
+            Bundler bundler = BundleHelper.BundleApp(fixture, out string singleFile, BundleOptions.BundleNativeBinaries);
```
Do we need to make sure the bundle actually has something to extract so that it creates/adds the extraction directory? By default, we won't have anything to extract on  Linux, right?

Commit:
Include mockcoreclr in the bundle

Commit:
Revert changes

Commit:
Remove repeated test

agocke commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/SingleFileApiTests.cs:
```diff
@@ -85,13 +85,52 @@ public void GetCommandLineArgs_0_Non_Bundled_App()
         }
 
+        [Fact]
+        public void AppContext_Native_Search_Dirs_Contains_Bundle_Dir()
```
I think you could integrate with the SelfContained tests above like follows:

```C#
        [Fact]
        public void SelfContained_SingleFile_APITests()
        {
            var fixture = sharedTestState.TestFixture.Copy();
            Bundler bundler = BundleHelper.BundleApp(fixture, out string singleFile);
            string extractionDir = BundleHelper.GetExtractionDir(fixture, bundler).Name;
            string bundleDir = BundleHelper.GetBundleDir(fixture).FullName;

            Command.Create(singleFile, "fullyqualifiedname codebase appcontext cmdlineargs executing_assembly_location basedirectory native_search_dirs")
                .CaptureStdErr()
                .CaptureStdOut()
                .Execute()
                .Should()
                .Pass()
                .And.HaveStdOutContaining("FullyQualifiedName: <Unknown>")
                .And.HaveStdOutContaining("Name: <Unknown>")
                .And.HaveStdOutContaining("CodeBase NotSupported")
                .And.NotHaveStdOutContaining("SingleFileApiTests.deps.json")
                .And.NotHaveStdOutContaining("Microsoft.NETCore.App.deps.json")
                // For single-file, Environment.GetCommandLineArgs[0] should return the file path of the host.
                .And.HaveStdOutContaining("Command line args: " + singleFile)
                .And.HaveStdOutContaining("ExecutingAssembly.Location: " + Environment.NewLine)
                .And.HaveStdOutContaining("AppContext.BaseDirectory: " + Path.GetDirectoryName(singleFile))
                // If we don't extract anything to disk, the extraction dir shouldn't
                // appear in the native search dirs.
                .And.HaveStdOutContaining($"NATIVE_DLL_SEARCH_DIRECTORIES: {bundleDir}:{Environment.NewLine}")
                .And.NotHaveStdOutContaining(extractionDir);
        }

        [Fact]
        public void SelfContained_NetCoreApp3_CompatMode_SingleFile_APITests()
        {
            var fixture = sharedTestState.TestFixture.Copy();
            Bundler bundler = BundleHelper.BundleApp(fixture, out string singleFile, BundleOptions.BundleAllContent);
            var extractionDir = BundleHelper.GetExtractionDir(fixture, bundler).FullName;
            string bundleDir = BundleHelper.GetBundleDir(fixture).FullName;
            var extractionBaseDir = BundleHelper.GetExtractionRootDir(fixture);

            Command.Create(singleFile, "fullyqualifiedname codebase appcontext cmdlineargs executing_assembly_location basedirectory native_search_dirs")
                .CaptureStdErr()
                .CaptureStdOut()
                .EnvironmentVariable(BundleHelper.DotnetBundleExtractBaseEnvVariable, extractionBaseDir.FullName)
                .Execute()
                .Should()
                .Pass()
                .And.HaveStdOutContaining(Path.DirectorySeparatorChar + "System.Private.CoreLib.dll") // In extraction directory
                .And.HaveStdOutContaining("System.Private.CoreLib.dll") // In extraction directory
                .And.NotHaveStdOutContaining("CodeBase NotSupported") // CodeBase should point to extraction directory
                .And.HaveStdOutContaining("SingleFileApiTests.dll")
                .And.HaveStdOutContaining("SingleFileApiTests.deps.json") // The app's .deps.json should be available
                .And.NotHaveStdOutContaining("Microsoft.NETCore.App.deps.json") // No framework - it's self-contained
                // For single-file, Environment.GetCommandLineArgs[0] should return the file path of the host.
                .And.HaveStdOutContaining("Command line args: " + singleFile)
                .And.HaveStdOutContaining("ExecutingAssembly.Location: " + extractionBaseDir.FullName) // Should point to the app's dll
                .And.HaveStdOutContaining("AppContext.BaseDirectory: " + extractionBaseDir.FullName) // Should point to the extraction directory
                // In extraction mode, we should have both dirs
                .And.HaveStdOutContaining($"NATIVE_DLL_SEARCH_DIRECTORIES: {extractionDir}:{bundleDir}:{Environment.NewLine}");
        }
```

agocke commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/SingleFileApiTests.cs:
```diff
@@ -85,13 +85,52 @@ public void GetCommandLineArgs_0_Non_Bundled_App()
+            // appear in the native search dirs.
+            Command.Create(singleFile, "native_search_dirs")
+                .CaptureStdErr()
+                .CaptureStdOut()
```
Should you be passing the `EnvironmentVariable(BundleHelper.DotnetBundleExtractBaseEnvVariable` stuff here?

agocke commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/SingleFileApiTests.cs:
```diff
@@ -85,13 +85,52 @@ public void GetCommandLineArgs_0_Non_Bundled_App()
+                .CaptureStdOut()
+                .Execute()
+                .Should().Pass()
+                .And.HaveStdOutContaining(bundleDir)
```
Any reason not to validate the whole string? Seems more robust. e.g.

```C#
.And.HaveStdOutContaining($"NATIVE_DLL_SEARCH_DIRECTORIES: {extractionDir}:{bundleDir}:{Environment.NewLine}");
```

agocke commented:
@mateoatr I don't think these changes are strictly necessary, they're more test hygiene suggestions. If you think it would be faster to deal with these separately, that's fine. We can merge this change, and take the modifications in another PR.

mateoatr commented:
@agocke I had some issues testing locally, but I just pushed your suggestions. I think we can merge after all tests run in the CI.

agocke commented:
Yeah, I've been trying to work on stuff locally and mostly got stuff to work by building everything, then runnign the apphost test by doing `~/runtime/dotnet.sh build` then `~/runtime/dotnet.sh build -t:test`

agocke commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/SingleFileApiTests.cs:
```diff
@@ -62,27 +69,10 @@ public void SelfContained_NetCoreApp3_CompatMode_SingleFile_APITests()
+                .And.HaveStdOutContaining("ExecutingAssembly.Location: " + extractionDir) // Should point to the app's dll
+                .And.HaveStdOutContaining("AppContext.BaseDirectory: " + extractionDir) // Should point to the extraction directory
+                // In extraction mode, we should have both dirs
+                .And.HaveStdOutMatching($"NATIVE_DLL_SEARCH_DIRECTORIES: .*{extractionDir}.*{bundleDir}");
```
Huh, I assume the `Matching` is turning this into a regex, meaning that the `\`s now need to be escaped on Windows?

agocke commented:
Moving us back to the previous commit, since that passed tests, so we can get this in today
